core/src/main/scala/cromwell/core/actor/BatchActor.scala

```diff
@@ -0,0 +1,127 @@
+package cromwell.core.actor
+
+import akka.actor.{ActorRef, FSM, Timers}
+import akka.dispatch.ControlMessage
+import common.collections.WeightedQueue
+import cromwell.core.actor.BatchActor._
+import cromwell.util.GracefulShutdownHelper.ShutdownCommand
+import org.slf4j.LoggerFactory
+
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+import scala.util.{Failure, Success}
+
+
+/** A collection of state, data, and message types to support BatchActor. */
+object BatchActor {
+  type BatchData[C] = WeightedQueue[C, Int]
+
+  val logger = LoggerFactory.getLogger("BatchBufferActor")
+
+  /** The states for BatchActor. */
+  sealed trait BatchActorState
+  case object WaitingToProcess extends BatchActorState
+  case object Processing extends BatchActorState
+
+  /** The control messages for BatchActor.
+    *  Because they extends ControlMessage, in combination with a control aware mailbox those messages will
+    *  be processed in priority.
+    */
+  sealed trait BatchActorControleMessage extends ControlMessage
+  case object ProcessingComplete extends BatchActorControleMessage
+  case object ScheduledFlushKey
+  case object ScheduledProcessAction extends BatchActorControleMessage
+  
+  case class CommandAndReplyTo[C](command: C, replyTo: ActorRef)
+
+  case class QueueWeight(weight: Int)
+}
+
+/**
+  * Abstract FSM that provides queueing and batching behavior based on a batch size and a flush rate.
+  * It is backed by a WeightedQueue which makes it possible to decouple the number of messages received from
+  * the effective "weight" of the queue.
+  */
+abstract class BatchActor[C](val flushRate: FiniteDuration,
+                             val batchSize: Int) extends FSM[BatchActorState, BatchData[C]] with Timers {
+  private var shuttingDown: Boolean = false
+  
+  implicit val ec = context.dispatcher
+  private val name = self.path.name
+
+  def isShuttingDown: Boolean = shuttingDown
+
+  override def preStart(): Unit = {
+    log.info("{} configured to flush with batch size {} and processHead rate {}.", name, batchSize, flushRate)
+    timers.startPeriodicTimer(ScheduledFlushKey, ScheduledProcessAction, flushRate)
+  }
+
+  startWith(WaitingToProcess, WeightedQueue.empty[C, Int](sizeFunction))
+
+  def commandToData(snd: ActorRef): PartialFunction[Any, C]
+  
+  when(WaitingToProcess) {
+    // On a regular event, only process if we're above the batch size is reached
+    case Event(command, data) if commandToData(sender).isDefinedAt(command) =>
+      processIfBatchSizedReached(data.enqueue(commandToData(sender)(command)))
+    // On a scheduled process, always process. Use the opportunity to broadcast the current queue weight
+    case Event(ScheduledProcessAction, data) =>
+      gossip(QueueWeight(data.weight))
```

Comment:
😮 😎

Reply:
does this data make me look fat?

Reply:
I'd only ever seen gossip in akka as being tied to cluster stuff. what is this?

Reply:
it's a built-in way simple eay to do event listening. You send a `Listen(your_actor)` to an FSM (or any actor extending the `Listeners` trait). It remembers you and when you call `gossip(message)` it sends the message to all registered listeners.